app/ptracker.py

```diff
@@ -143,5 +143,42 @@ def get_pageloads_per_company():
     return jsonify(response_body)
 
 
+@app.route("/read/<rule_id>", methods=["GET"])
+def reached_end_of_page(rule_id):
+    Session = sessionmaker(bind=engine)
+    session = Session()
+
+    record = None
+
+    try:
+        record = session.query(Pageloads).get(rule_id)
+    except Exception as e:
+        session.close()
+        return Response(status=500, response="Unknown error occured")
+
+    if record is None:
+        session.close()
+        return Response(status=404)
+
+    # if date is already filled, do not update
+    if record.page_end is not None:
+        session.close()
+        return Response(status=400)
+
+    record.page_end = datetime.utcnow()
+    session.add(record)
```

Comment:
I am pretty sure that this line can be removed.  My reading of the [documentation](https://docs.sqlalchemy.org/en/13/orm/session_api.html#sqlalchemy.orm.session.Session.add) seems to indicate that it should silently fail due to double adding an object.  

Having said all that the code does work

Reply:
From what I understand `session.add` is necessary. Code seems to work as it is so i will leave it as is and merge